src/v/pandaproxy/schema_registry/seq_writer.h

```diff
@@ -0,0 +1,115 @@
+//// Copyright 2021 Vectorized, Inc.
+////
+//// Use of this software is governed by the Business Source License
+//// included in the file licenses/BSL.md
+////
+//// As of the Change Date specified in that file, in accordance with
+//// the Business Source License, use of this software will be governed
+//// by the Apache License, Version 2.
+
+#pragma once
+
+#include "kafka/client/client.h"
+#include "pandaproxy/schema_registry/sharded_store.h"
+#include "pandaproxy/schema_registry/types.h"
+#include "random/simple_time_jitter.h"
+#include "utils/retry.h"
+
+namespace pandaproxy::schema_registry {
+
+using namespace std::chrono_literals;
+
+static const int max_retries = 32;
+
+class seq_writer final : public ss::peering_sharded_service<seq_writer> {
+public:
+    seq_writer(
+      ss::smp_service_group smp_group,
+      ss::sharded<kafka::client::client>& client,
+      sharded_store& store)
+      : _smp_opts(ss::smp_submit_to_options{smp_group})
+      , _client(client)
+      , _store(store) {}
+
+    ss::future<> read_sync();
+
+    // API for readers: notify us when they have read and applied an offset
+    ss::future<> advance_offset(model::offset offset);
+
+private:
+    ss::smp_submit_to_options _smp_opts;
+
+    ss::sharded<kafka::client::client>& _client;
+    sharded_store& _store;
+
+    void advance_offset_inner(model::offset offset);
+
+    struct WriteCollision : public std::exception {
+        const char* what() const throw() { return "Write Collision"; }
+    };
+
+    simple_time_jitter<ss::lowres_clock> _jitter{
+      std::chrono::milliseconds{100}};
+
+    /// Helper for write paths that use sequence+retry logic to synchronize
+    /// multiple writing nodes.
+    template<typename F>
+    auto sequenced_write(F f) {
+        auto base_backoff = _jitter.next_duration();
+        auto remote = [base_backoff, f](seq_writer& seq) {
+            auto units = ss::get_units(seq._write_sem, 1);
```

Comment:
Are these units supposed to be held across the retry? Is so, they should be moved to a finally continuation after the `retry_with_backoff`.

Reply:
As elsewhere -- replaced with with_semaphore.